### ISSUE TITLE
Vulkan: Can't start Godot with AMD integrated + NVIDIA dedicated GPU switchable graphics (Failed dispatch call `vkEnumeratePhysicalDevices`)

### ISSUE THREAD

**Issue by fonixgm** [godotengine/godot#57708]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0

### System information

Windows 10 RTX 2060 and AMD graphics chipset

### Issue description

![67c643612afe3503f692c94426fc251b](https://user-images.githubusercontent.com/40607420/152685307-21b277c9-aabd-4294-a42d-60dcc8bcb8de.png)
![b09673e306d981f59a5e5606dd42dec6](https://user-images.githubusercontent.com/40607420/152685566-7be821d1-6fa5-41ff-a56d-231e4c84f99c.png)

When I double click on the program I get the following error showing that I don't have the drivers I have checked everything and it is correct I have even installed it again and it is still the same.

### Steps to reproduce

Open ./Godot_v4.0-alpha1_win64_console.cmd or ./Godot_v4.0-alpha1_win64.exe

**Comment by fonixgm** [comment]
The solution I had to apply is the following by performing these actions from this video:
https://www.youtube.com/watch?v=0HdHE_sq7dw

**Comment by Calinou** [comment]
I don't think there's anything we can do about this. Setting `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` to `1` automatically is probably not a good idea.

cc @bruvzg

**Comment by fonixgm** [comment]
> Related to #57396.
> 
> I don't think there's anything we can do about this. Setting `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` to `1` automatically is probably not a good idea, as it will prevent forcing the use of AMD integrated graphics entirely when a dedicated GPU is present. For non-game applications, this can be problematic if you want to preserve battery life.
> 
> Did you update your graphics drivers from AMD and NVIDIA's websites (**not** from your laptop OEM or Windows Update)?
> 
> cc @bruvzg

drivers are updated from windows update as well as from nvidia and amd chipset manufacturers.

**Comment by bruvzg** [comment]
> as it will prevent forcing the use of AMD integrated graphics entirely when a dedicated GPU is present

If I understand correctly, with `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` both devices should be accessible, without it, it's presented as a single device.

@fonixgm Could you try running Godot with `--verbose` command line argument (both with and without `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` set), and give us the output ("Vulkan devices:" part in particular).

**Comment by fonixgm** [comment]
> > as it will prevent forcing the use of AMD integrated graphics entirely when a dedicated GPU is present
> 
> If I understand correctly, with `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` both devices should be accessible, without it, it's presented as a single device.
> 
> @fonixgm Could you try running Godot with `--verbose` command line argument (both with and without `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` set), and give us the output ("Vulkan devices:" part in particular).

This is with DISABLE_LAYER_AMD_SWITCHCHABLE_GRAPHICS_1 on
![16db03304dd4429b3e5ccb0e52725cee](https://user-images.githubusercontent.com/40607420/152686231-f30e035f-37ed-432a-8fe5-12bcc33aa906.png)
![8e3662aaaa748a204234a14de0fa0b63](https://user-images.githubusercontent.com/40607420/152686237-7eb70fc5-d55c-423f-a77a-51cc90ad9277.png)

and this is without the DISABLE_LAYER_AMD_SWITCHCHABLE_GRAPHICS_1 
It seems that when you turn it on by command console it is working and boots normally without having this option activated, it is a little strange.

![fae39c1b3131c350ff4f8b5d46915243](https://user-images.githubusercontent.com/40607420/152686299-a376bb56-01de-49ab-8fe4-29db1ea84fa0.png)
![5b851beb6b3cf23618422d69b7fb9db6](https://user-images.githubusercontent.com/40607420/152686344-a53f85f9-6b8a-4a4f-94ae-a7e5536ea14d.png)

**Comment by bruvzg** [comment]
I have not found much useful info how `VK_LAYER_AMD_SWITCHCHABLE_GRAPHICS` should work, but it seems broken (there are also plenty of similar issues with the layer on Linux, so it's not Windows specific).

With the layer disabled, it should be possible to force a specific GPU for the app using Windows built-in option in "System→Display→Graphics". But I agree that we probably can't do anything with it.

**Comment by fonixgm** [comment]
> I have not found much useful info how `VK_LAYER_AMD_SWITCHCHABLE_GRAPHICS` should work, but it seems broken (there are also plenty of similar issues with the layer on Linux, so it's not Windows specific).
> 
> With the layer disabled, it should be possible to force a specific GPU for the app using Windows built-in option in "System→Display→Graphics". But I agree that we probably can't do anything with it.

Temporarily I will use the DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1 to make it work and I will also investigate if there is any configuration that can solve this problem apart from using this environment variable and thus not to annoy so much the battery of my pc.

**Comment by ChristianB84** [comment]
I reproduced this with a similar combination of graphics cards and fixed it locally by adding "OS::get_singleton()->set_environment("DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1", "1");" to the function Main::setup in main.cpp, immediately after "OS::get_singleton()->initialize();". At least the Windows implementation for that function changes the environment variable only for the current process (at least according to the documentation). Is there any reason not to do this? Updating my graphics drivers didn't do the trick and I guess it might take some time till most users with such a (not that unusual) graphics cards combination will receive an update that fixes this issue.

**Comment by Mikadini** [comment]
i want to comunicate i solve a similar problem ( #56164 ) i simply remove the official asus amd driver and put the generale amd driver using the adrenalin software this solve the problem for me.

**Comment by Calinou** [comment]
> I reproduced this with a similar combination of graphics cards and fixed it locally by adding "OS::get_singleton()->set_environment("DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1", "1");" to the function Main::setup in main.cpp, immediately after "OS::get_singleton()->initialize();". At least the Windows implementation for that function changes the environment variable only for the current process (at least according to the documentation). Is there any reason not to do this? Updating my graphics drivers didn't do the trick and I guess it might take some time till most users with such a (not that unusual) graphics cards combination will receive an update that fixes this issue.

How will this behave on systems with *only* AMD integrated graphics?

**Comment by akien-mga** [comment]
Please try 4.0 alpha 10, there was an update to VMA which might help: https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-10

**Comment by fonixgm** [comment]
I have tried alpha 10 but the error persists.

![image](https://user-images.githubusercontent.com/40607420/174284806-d6cba514-be2b-4cc5-ba38-bf843f6953ee.png)

**Comment by Ramedes** [comment]
The Error occurs on Branch 3.x, 3.4.x and 4.x

On my System i have a integrated Radeon AMD Vegas 11 Graphics card and a additional, second NVidia Graphics card.
I simply turned the AMD Vergas Graphic card up in Windows 10 Device manager (Nvidia still was enabled), the Editor
launched without any problems. After it, i was able to re-enable the AMD Graphics card in Device Manager (without Reboot)
and the Editor starts without any Problems.

**Comment by Calinou** [comment]
> The Error occurs on Branch 3.x, 3.4.x and 4.x
> 
> On my System i have a integrated Radeon AMD Vegas 11 Graphics card and a additional, second NVidia Graphics card. I simply turned the AMD Vergas Graphic card up in Windows 10 Device manager (Nvidia still was enabled), the Editor launched without any problems. After it, i was able to re-enable the AMD Graphics card in Device Manager (without Reboot) and the Editor starts without any Problems.

Godot 3.x uses an OpenGL renderer, so Vulkan-specific issues don't affect it.

**Comment by Ramedes** [comment]
I built 3.x, 3.4 and 4.x and the error shows up on any build, with tools or without, of it is release or debug targets or both.

**Comment by ChristianB84** [comment]
> > I reproduced this with a similar combination of graphics cards and fixed it locally by adding "OS::get_singleton()->set_environment("DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1", "1");" to the function Main::setup in main.cpp, immediately after "OS::get_singleton()->initialize();". At least the Windows implementation for that function changes the environment variable only for the current process (at least according to the documentation). Is there any reason not to do this? Updating my graphics drivers didn't do the trick and I guess it might take some time till most users with such a (not that unusual) graphics cards combination will receive an update that fixes this issue.
> 
> How will this behave on systems with _only_ AMD integrated graphics? Can anyone test this on such a system?

I don't know. I don't have such a system. Can anyone with an AMD-only system test it?

**Comment by fonixgm** [comment]
I have done an update of the latest driviers of the radeon integrated graphics and the amd just install the necessary drivers without doing a minimal installation nor a full installation when I have done this suddenly I have noticed an improvement and I am not failing the godot or shows any error when it boots it does it directly from the rtx.

I give for my part solved the problem I think it was probolema of amd and its drivers.

Translated with www.DeepL.com/Translator (free version)

**Comment by fonixgm** [comment]
if someone fails to check this as shown on screen in drivers only and this should not fail or at least it does not fail me
![Screenshot 2022-08-07 001403](https://user-images.githubusercontent.com/40607420/183267638-1dd818ca-f731-439c-8e51-a6df74b39f36.png)

**Comment by ChristianB84** [comment]
The main problem here isn't whether or not this works for developers. There are workarounds. The important question is: will someone who wants to play (or even buy) a game made with Godot run into this issue? Will they be willing to take any additional steps to make that one particular game work? Keep in mind that making changes to one's setup like setting environment variables isn't something everyone is familiar with and updating a graphics driver can even cause additional issues if the new version is bugged in any way.

I didn't have this problem with anything but Godot so far. So if I had first encountered this issue with a game instead, I'd blame it on the game. And I guess that's what others will do as well. And they might write angry reviews that will stop potential players (or customers) from buying the game.

**Comment by Calinou** [comment]
> The important question is: will someone who wants to play (or even buy) a game made with Godot run into this issue?

Yes, as this issue affects both the editor and exported projects. We don't know a definite way to fix this, and contributors work on a best-effort basis.

**Comment by ChristianB84** [comment]
I know. The point of my post was just to make sure this issue isn't closed accidently as being solved. I agree that it isn't really urgent to fix it.

**Comment by lmReef** [comment]
> I have done an update of the latest driviers of the radeon integrated graphics and the amd just install the necessary drivers without doing a minimal installation nor a full installation when I have done this suddenly I have noticed an improvement and I am not failing the godot or shows any error when it boots it does it directly from the rtx.
> 
> I give for my part solved the problem I think it was probolema of amd and its drivers.

Installing the latest AMD drivers fixed this for me as well.
Windows 11, AMD Ryzen 7 5800H with Radeon Graphics, NVIDIA GeForce RTX 3060.

**Comment by arocull** [comment]
Bumping because I had this issue too.
An easy, temporary solution on Windows to force Godot to use the NVIDIA GPU is editing the cmd script to locally set the `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` environment variable only for the given session.
```cmd
@echo off
setlocal
	set DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1=1
	Godot_v4.0-beta1_win64.exe
	pause > nul
endlocal
```

**Comment by Redwarx008** [comment]
Update AMD Radeon diver can solve this problem

**Comment by Calinou** [comment]
If any of you can still reproduce the issue, please test https://github.com/godotengine/godot/pull/73450 locally. A Windows build is included in that PR's description.

**Comment by MonT97** [comment]
> Update AMD Radeon diver can solve this problem

confirmed here :
R7 4800H
GTX 1650

got the driver from AMD website

**Comment by abderman** [comment]
> The solution I had to apply is the following by performing these actions from this video: https://www.youtube.com/watch?v=0HdHE_sq7dw

Thank you, its'working!!!

**Comment by yarre-uk** [comment]
> Bumping because I had this issue too. An easy, temporary solution on Windows to force Godot to use the NVIDIA GPU is editing the cmd script to locally set the `DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1` environment variable only for the given session.
> 
> ```batchfile
> @echo off
> setlocal
> 	set DISABLE_LAYER_AMD_SWITCHABLE_GRAPHICS_1=1
> 	Godot_v4.0-beta1_win64.exe
> 	pause > nul
> endlocal
> ```

hello, tried your fix, and godot closes without any errors, just "not responding"

**Comment by snailrhymer** [comment]
> If any of you can still reproduce the issue, please test #73450 locally. A Windows build is included in that PR's description.

Works for me!

**Comment by TATAPNHOB** [comment]
![image](https://github.com/godotengine/godot/assets/78953252/c9cb4c9f-0ebf-43e1-95ec-e3b22cf365c2)
![image](https://github.com/godotengine/godot/assets/78953252/bff89fd3-1c37-4eed-b9ef-87c661a6672b)
My errors are like these. It seems that it recognizes the dedicated GPU, but still cannot initialize Vulkan.

**Comment by Calinou** [comment]
> My errors are like these. It seems that it recognizes the dedicated GPU, but still cannot initialize Vulkan. Actions from the aforementioned video do not work, as it seems.

Try [4.1.rc2](https://godotengine.org/article/release-candidate-godot-4-1-rc-2/) as it includes https://github.com/godotengine/godot/pull/73450.

**Comment by TATAPNHOB** [comment]
4.1.rc2 works! Thank you. I don't have any external monitors though.

**Comment by fudgepop01** [comment]
for those of you arriving here in 2024 after January 27th, the issue is detailed here:
https://community.khronos.org/t/cannot-create-vulkan-instance-non-conformant-vulkan-implementation/110488/3

> I’m having the same problem. Yesterday, I could open Vulkan-supported applications, but now I can’t. I updated the driver, used DDU to uninstall and reinstall it, but nothing changed. Example Vulkan applications also don’t open. I downloaded “VulkanRT-1.3.275.0-Components” from the LunarG website. The x86 version works with a warning that “dzn is not a conformant Vulkan implementation, testing use only.” However, the x64 version only gives a warning. Sometimes GPU-Z shows that it supports Vulkan, but then it disappears when the application is refreshed. My system: R5 5600G Vega 7(Cezzane), OS: Windows 11.

> **Edit: Uninstalling the “OpenCL™, OpenGL® and Vulkan® Compatibility Pack” fixed the problem.**

This solution worked for me. You can easily find this in "Apps > Installed Apps" in Windows settings.

**Comment by nathan130200** [comment]
bro... this is not fixed yet?

**Comment by lmReef** [comment]
> bro... this is not fixed yet?

I've stopped running into this issue some time ago using godot 4. Though I'm not sure exactly which version fixed for me I'm currently on 4.2.1

**Comment by akien-mga** [comment]
This was fixed in 4.1. If you still experience a similar issue in 4.2.1 or later, please open a new issue with details.
